### PR TITLE
Implement unpivot in cudf-polars

### PR DESCRIPTION
**Pull request by wence-**
## Description

Add support for unpivoting a DataFrame. We raise for cases where the concatenation of the value columns produces a cast that is not supported by standard fixed-width unary casting.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by wence-** [comment]
This needs to handle defaulted values for the to-pivot and index columns

**Comment by wence-** [comment]
> This needs to handle defaulted values for the to-pivot and index columns

Done.

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 1]
```diff
@@ -1115,6 +1129,43 @@ def evaluate(self, *, cache: MutableMapping[int, DataFrame]) -> DataFrame:
             return DataFrame.from_table(
                 plc.lists.explode_outer(df.table, index), df.column_names
             ).sorted_like(df, subset=subset)
+        elif self.name == "unpivot":
+            indices, pivotees, variable_name, value_name = self.options
+            npiv = len(pivotees)
+            df = self.df.evaluate(cache=cache)
+            if indices:
```
Comment (vyasr) [inline python/cudf_polars/cudf_polars/dsl/ir.py:1136]: Does indices default to None or to an empty sequence? If it's the latter, the `if` block would just work without needing a separate else IIUC. If it doesn't, that seems like an edge case bug to fix in plc. If it's the former, then nothing to do here.

File: python/cudf_polars/tests/test_mapfunction.py [hunk 2]
```diff
@@ -61,3 +61,48 @@ def test_rename_columns(mapping):
...
+        }
+    )
+    q = df.unpivot(
+        ["c", "b"], index=index, variable_name=variable_name, value_name=value_name
+    )
+
+    assert_gpu_result_equal(q)
+
+
+def test_unpivot_defaults():
```
Comment (vyasr) [inline python/cudf_polars/tests/test_mapfunction.py:85]: Wouldn't this test already be covered by the None parametrizations in the test above if the `on` argument was parametrized with None in addition to `["c", "b"]`?

File: python/cudf_polars/tests/test_mapfunction.py [hunk 3]
```diff
@@ -61,3 +61,48 @@ def test_rename_columns(mapping):
...
+    df = pl.LazyFrame(
+        {
+            "a": ["x", "y", "z"],
+            "b": pl.Series([1, 3, 5], dtype=pl.Int16),
+        }
+    )
+
+    q = df.unpivot(["a", "b"])
+
+    assert_ir_translation_raises(q, NotImplementedError)
```
Comment (vyasr) [inline python/cudf_polars/tests/test_mapfunction.py:108]: Out of scope for this PR, but in a follow-up we should make `NotImplementedError` the default for this function. It is almost always the failure that we expect when translation fails.

File: python/cudf_polars/cudf_polars/containers/column.py [hunk 4]
```diff
@@ -84,6 +84,34 @@ def sorted_like(self, like: Column, /) -> Self:
...
 
+    # TODO: Return Column once #16272 is fixed.
+    def astype(self, dtype: plc.DataType) -> plc.Column:
+        """
+        Return the backing column as the requested dtype.
+
+        Parameters
+        ----------
+        dtype
+            Datatype to cast to.
```
[the remark this answers is not in the document]
Comment (wence-) [reply python/cudf_polars/cudf_polars/containers/column.py:95]: No.

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 5]
```diff
@@ -1094,6 +1095,19 @@ def __post_init__(self) -> None:
...
+            value_name = "value" if value_name is None else value_name
+            variable_name = "variable" if variable_name is None else variable_name
+            if len(pivotees) == 0:
+                index = frozenset(indices)
+                pivotees = [name for name in self.df.schema if name not in index]
+            if not all(
+                dtypes.can_cast(self.df.schema[p], self.schema[value_name])
+                for p in pivotees
+            ):
+                raise NotImplementedError("unpivot with non-fixed-width columns.")
```
[the remark this answers is not in the document]
Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/ir.py:1109]: I think a more general error message is the right thing. "Cannot produce column of type bla from type foo"

File: python/cudf_polars/cudf_polars/dsl/ir.py (hunk 1 above)
Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/ir.py:1136]: Empty sequence, so yes, I can simplify. _but_, does `cudf::reshape` work right with an empty table? I will check.

File: python/cudf_polars/tests/test_mapfunction.py (hunk 2 above)
Comment (wence-) [reply python/cudf_polars/tests/test_mapfunction.py:85]: Yes, but then we would have some failing tests because we can't unpivot columns of strings and numbers together.

Comment (vyasr) [reply python/cudf_polars/tests/test_mapfunction.py:85]: Ah OK got it. Then let's leave it as is.

**Review by vyasr — APPROVED** [review]
We agree on the desired resolutions to the open threads, so approving assuming that they are addressed (the change to the assertion function, if desired, would be done in a follow-up anyway so I'm OK leaving that thread open and unaddressed for now).

File: python/cudf_polars/cudf_polars/containers/column.py [hunk 6]
```diff
@@ -84,6 +84,34 @@ def sorted_like(self, like: Column, /) -> Self:
...
+        RuntimeError
+            If the cast is unsupported.
+
+        Notes
+        -----
+        This only produces a copy if the requested dtype doesn't match
+        the current one.
+        """
+        if self.obj.type() != dtype:
+            return plc.unary.cast(self.obj, dtype)
```
[the remark this answers is not in the document]
Comment (wence-) [reply python/cudf_polars/cudf_polars/containers/column.py:112]: I think there's a misunderstanding here. `astype` turns a `Column` with dtype `X` into one with dtype `Y`. If `Y == X`, we don't do anything. If `Y != X` then we must do something.

The _precondition_ for calling `astype` right now (because we haven't built out casting utilities) is that `dtypes.can_cast` returns true. So effectively we have:

```
if can_cast(source, target):
    value = column.astype(target) # always succeeds, maybe copies.
```

So I don't think the condition will change.

File: python/cudf_polars/cudf_polars/dsl/ir.py (hunk 1 above)
Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/ir.py:1136]: Done.

File: python/cudf_polars/cudf_polars/dsl/ir.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/ir.py]: Done.

File: python/cudf_polars/cudf_polars/containers/column.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf_polars/cudf_polars/containers/column.py:112]: > The precondition for calling astype right now (because we haven't built out casting utilities) is that dtypes.can_cast returns true.

I didn't realize this until later in the review, hence my confusion. I forgot to come back to this thread. Carry on, nothing to see here.
